### PR TITLE
nixos/simplesamlphp: init

### PR DESCRIPTION
**Pull request by thenhnn**
## Description of changes

Packaging of [SimpleSAMLphp](https://simplesamlphp.org/), an application written in native PHP that deals with authentication.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by thenhnn** [comment]
@drupol, can you check this one?

File: pkgs/by-name/si/simplesamlphp/package.nix [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  version = "1.19.7";
+
+  src = fetchFromGitHub {
+    owner = "simplesamlphp";
+    repo = "simplesamlphp";
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-Qmy9fuZq8MBqvYV6/u3Dg92pHHicuUhdNeB22u4hwwA=";
+  };
+
+  php = php.buildEnv {
```
Comment (drupol) [inline pkgs/by-name/si/simplesamlphp/package.nix:16]: Why do you need a custom version of PHP ?

**Review by drupol — CHANGES_REQUESTED** [review]
Could you please pass the new files you're introducing through `nixfmt` (`nixfmt-rfc-style`) ?

File: pkgs/by-name/si/simplesamlphp/package.nix (hunk 1 above)
Comment (thenhnn) [reply pkgs/by-name/si/simplesamlphp/package.nix:16]: SimpleSAMLphp requires PHP extensions, and as I understand, it requires custom version of PHP.

Comment (drupol) [reply pkgs/by-name/si/simplesamlphp/package.nix:16]: The default PHP version already includes a comprehensive set of extensions, see: https://github.com/NixOS/nixpkgs/blob/d5f39744c87dbbfe55a1d8ae2da409293b2d2760/pkgs/development/interpreters/php/8.1.nix#L17

Isn't this enough?

**Comment by thenhnn** [comment]
> Could you please pass the new files you're introducing through `nixfmt` (`nixfmt-rfc-style`) ?

Thanks, I didn't know about `nixfmt-rfc-style` :)

File: pkgs/by-name/si/simplesamlphp/package.nix (hunk 1 above)
Comment (thenhnn) [reply pkgs/by-name/si/simplesamlphp/package.nix:17]: or I am wrong?

**Comment by drupol** [comment]
> > Could you please pass the new files you're introducing through `nixfmt` (`nixfmt-rfc-style`) ?
> 
> Thanks, I didn't know about `nixfmt-rfc-style` :)

More info in https://github.com/NixOS/rfcs/pull/166

File: pkgs/by-name/si/simplesamlphp/package.nix (hunk 1 above)
Comment (drupol) [reply pkgs/by-name/si/simplesamlphp/package.nix:17]: I'm asking you...

Comment (thenhnn) [reply pkgs/by-name/si/simplesamlphp/package.nix:17]: Oh, that's enough. I will delete phpEnv then.

Comment (drupol) [reply pkgs/by-name/si/simplesamlphp/package.nix:17]: I think you can delete the `php` attribute indeed.

Comment (thenhnn) [reply pkgs/by-name/si/simplesamlphp/package.nix]: I deleted it. Thank you!

File: nixos/modules/services/web-apps/simplesamlphp.nix [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  config = {
+    services.phpfpm.pools = lib.mapAttrs' (
+      phpfpmName: opts:
+      lib.nameValuePair opts.phpfpmPool { phpEnv.SIMPLESAMLPHP_CONFIG_DIR = "${generateConfig opts}"; }
+    ) cfg;
+
+    services.nginx.virtualHosts = lib.mapAttrs' (
+      phpfpmName: opts:
+      let
```
Comment (drupol) [inline nixos/modules/services/web-apps/simplesamlphp.nix:108]: Please remove unneeded `let ... in` clause.

Comment (thenhnn) [reply nixos/modules/services/web-apps/simplesamlphp.nix]: oops

**Comment by drupol** [comment]
Hi,

No worries, you don't annoy me.

Have you tested the module on your own, is it working?

**Comment by thenhnn** [comment]
> Hi,
> 
> No worries, you don't annoy me.
> 
> Have you tested the module on your own, is it working?

Yeah, it is working with filesender (ongoing pr that I will undraft or recreate when this PR will be merged).

**Comment by thenhnn** [comment]
Btw, should I mention somewhere in docs that this module is supposed to work with pair with some PHP service?
"PHP library for authentication" can be somewhat confusing probably. I'm not sure tho.

**Comment by drupol** [comment]
> Btw, should I mention somewhere in docs that this module is supposed to work with pair with some PHP service? 

Could you please elaborate further?

**Comment by thenhnn** [comment]
> > Btw, should I mention somewhere in docs that this module is supposed to work with pair with some PHP service?
> 
> Could you please elaborate further?

SimpleSAMLphp by it's own is just a library. It cannot do anything useful except authentication of users. (It can be OAuth2 provider, but I seen issues with that + nixpkgs has way more better alternatives for that use case)

For example [filesender](https://github.com/filesender) uses it.

Maybe it's worth mentioning that main use-case of this module is pairing with other PHP services?

**Comment by drupol** [comment]
Yes I think it's a good idea to mention as much documentation as possible.

**Comment by thenhnn** [comment]
> Yes I think it's a good idea to mention as much documentation as possible.

Where I should add it? In module description?

**Comment by drupol** [comment]
Please do as you prefer, I don't have a strong opinion on this.

**Comment by thenhnn** [comment]
I added mention in module's description.

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 3]
```diff
@@ -217,6 +217,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - [keto](https://www.ory.sh/keto/), a permission & access control server, the first open source implementation of ["Zanzibar: Google's Consistent, Global Authorization System"](https://research.google/pubs/zanzibar-googles-consistent-global-authorization-system/).
 
+- [SimpleSAMLphp](https://simplesamlphp.org/), an application written in native PHP that deals with authentication. Available as [services.simplesamlphp](#opt-services.simplesamlphp).
```
[the remark this answers is not in the document]
Comment (thenhnn) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: Good idea, done.

**Comment by drupol** [comment]
Could you squash the two last commit? There's no need to have `@thenhnn
[nixos/doc: add simplesamlphp module mention](https://github.com/NixOS/nixpkgs/pull/312251/commits/a20f21c23ad47cc14bd2b766a1bcde101e38a592)` and it can be merged in `[nixos/simplesamlphp: init module](https://github.com/NixOS/nixpkgs/pull/312251/commits/fe1a558b65ffee9bc8139d6b45a48ae3a75e72b7)`

**Comment by thenhnn** [comment]
I squashed commits.

**Review by drupol — APPROVED** [review]

**Comment by thenhnn** [comment]
Should we ping someone else or one approval is enough?

**Comment by drupol** [comment]
I'm waiting for the CI to be done to merge it.

**Comment by thenhnn** [comment]
Thank you! :3

**Comment by Ma27** [comment]
I'm a little shocked, stumbling upon this.

This is two patch-levels __behind__ 1.19.9, the final version of 1.x which was released two months __before__ this PR was opened: https://github.com/simplesamlphp/simplesamlphp/releases/tag/v1.19.9

For a security-sensitive piece of software, this is frankly careless.

**Pull request merged**